working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+    {Object? exceptionalReturn});
+}
+```
+
+The function pointer returned by `NativeCallable.shared(...).nativeFunction`
+will be bound to an isolate group which produced it using the same trampoline
+mechanism FFI uses to create function pointers from closures. Returned function
+pointer can be called by native code from any thread. It does not require
+exclusive access to a specific isolate and thus avoids interoperability pitfalls
+associated with that:
+
+- No need to block native caller and wait for the target isolate to become
+  available.
+- Clear semantics of globals:
+  - `shared` global state is accessible and independent from the current thread;
+  - accessing non-`shared` state will throw an `IsolationError`.
+
+#### Shared Libraries created from Dart code
+
+An introduction of _shared isolate_ allows us to adjust our deployment story and
+make it simpler to create and use shared libraries from Dart code.
+
+Consider for example previously given in the
+[Interoperability](#interoperability) section:
+
+```dart
+// foo.dart
+
+import 'dart:ffi';
+
+@ffi.Export()
+void foo() {
+
+}
+```
+
+which produces a shared library exporting a C symbol:
+
+```cpp
+// foo.h
+
+extern "C" void foo();
+```
+
+Shared isolates give us a tool to define what happens when `foo` is invoked by a
+native caller:
+
+- There is a 1-1 correspondence between loaded shared library and an isolate
+  group corresponding to this shared library. This isolate group is created when
+  shared library is loaded (or possibly via explicit call to a specific exported
+  symbol) and is destroyed when shared library is unloaded (or possibly via a
+  call to specific symbol).
+- When an exported symbol is invoked the call happens in the shared isolate of
+  that isolate group.
+
+## Core Library Changes
+
+### Upgrading `dart:async` capabilities
+
+#### Shareable `Future` and `Stream` instances
+
+`Future` and `Stream` should receive the same treatment as `List`: `dart:async`
+should be extended with shareable versions of these and a way to convert an
+existing object to a shareable one.
+
+```dart
+class ShareableFuture<T implements Shareable?>
+    implements Future<T>, Shareable {
+}
+
+class ShareableStream<T implements Shareable?>
+    implements Stream<T>, Shareable {
+}
+
+extension ToShareableFuture<E extends Shareable?> on Future<E> {
+  /// Converts the [Future] to [ShareableFuture] if it is not already shareable.
+  ShareableFuture<E> toShareable() => /* ... */;
+}
+
+extension ToShareableStream<E extends Shareable?> on Stream<E> {
+  /// Converts the [Future] to [ShareableFuture] if it is not already shareable.
+  ShareableStream<E> toShareable() => /* ... */;
+}
+```
+
+The reason for providing these is to allow developers to structure their code
+using well understood primitives futures and streams instead of devising new
+primitives which reimplement some of the same functionality and is compatible
+with threading.
```

Comment:
A shared future can be completed with a shareable value, and that will be visible from every isolate where the future is reachable.

I assume callbacks registered on a shareable future are isolate specific?
And everything isolate which has a listener will need to register some callback which gets notified, in the isolate itself, that the future has completed.

This sounds like we'll need a cross isolate event/notification primitive.
An event that any code in any isolate can listen for, and when triggered, from anywhere using a shareable trigger, every listener will get notified in their own isolate.
If we have that, I think I can build shareable futures on top. No need to send values when we already have shared state, just a notification is enough.

Then there is the issue of errors, which aren't necessarily shareable. A shareable future may choose to only send the `toString` of the original error, if the error isn't sendable, and maybe its  type. Can make carding specific errors harder, though. Can't just parallelize computations of error handling isn't the same.

Streams have pushback too. Still, notification plus shared date should be enough primitives to implement that.